File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int result = 0;
        int min = Integer.MAX_VALUE;
        for(int i = 0; i < prices.length; i++) {
            if (prices[i] < min) {
                min = prices[i];
            }
            else{
                result = Math.max(result, prices[i]-min);
            }
        }
        return result;
        
    }
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = [[[0,float('-inf')]]*2]*(len(prices)+1)
        # print(dp)
        for i in range(len(prices)):
            dp[i+1][1][0] = max(dp[i][1][0], dp[i][1][1]+prices[i])
            dp[i+1][1][1] = max(dp[i][1][1], 0-prices[i])
        # print(dp)
        return dp[-1][1][0]